pkg/podcgroup/podcgroup.go

```diff
@@ -0,0 +1,84 @@
+package podcgroup
+
+import (
+	"io/ioutil"
+	"regexp"
+	"strings"
+)
+
+// ExtractIdFromCgroupPath finds the pod uid and container id from a cgroup
+// path.
+//
+// Examples of paths that can be parsed:
+// /sys/fs/cgroup/systemd/kubepods/besteffort/pod91a8fc3a-0ecf-48b4-81bf-78a7275d348c/f75aff467357c5d0ddd47cb7ad87ed38746e018992586ff66198a5c11218f634
+// /sys/fs/cgroup/systemd/kubepods.slice/kubepods-besteffort.slice/kubepods-besteffort-pod5759a1ae_36ca_48a8_b20a_b0b5c8a90fb8.slice/docker-c8b38413c88eefe063b8cd3f01c16be5e3bda9693a19a68a88807baca9feb937.scope
+// /sys/fs/cgroup/systemd/kubepods.slice/kubepods-burstable.slice/kubepods-burstable-pod12345678_1234_1234_1234_123456789012.slice/crio-f75aff467357c5d0ddd47cb7ad87ed38746e018992586ff66198a5c11218f634.scope
+func ExtractIdFromCgroupPath(cgroupPath string) (podUid, containerID string) {
+	if cgroupPath == "" {
```

Comment:
Another way could be something like:

```go
if !strings.HasPrefix(cgroupPath, "/sys/fs/cgroup/systemd/kubepods") {
	return
}
```

This would make the regexes below a bit shorter.

Reply:
Also, it looks like the interesting data is in last two segments of the path, so maybe:

```go
dir, containerIDSegment := filepath.Split(cgroupPath)
podIDSegment := filepath.Base(dir)
```

and then do the regexp matches on those parts?

```go
// idx 0 is whole match, idx 1 is the ID (potentially with underlines instead of dashes)
podIDRegexp := regexp.MustCompile(`pod([a-f0-9]{8}[-_][a-f0-9]{4}[-_][a-f0-9]{4}[-_][a-f0-9]{4}[-_][a-f0-9]{12})`)
// the docker- or crio- prefix is optional,
// idx 0 is whole match, idx 1 is the prefix without dash or empty if none, idx 2 is the ID
containerIDRegexp := regexp.MustCompile(`(?:([a-z]*)-)?([0-9a-f]{64})`)
```

Reply:
> `if !strings.HasPrefix(cgroupPath, "/sys/fs/cgroup/systemd/kubepods") {`

I can use `strings.HasPrefix` but I need to adapt this a bit.

The kubelet could be started with the `--cgroup-root` option ([documentation](https://kubernetes.io/docs/reference/command-line-tools-reference/kubelet/)):
```
--cgroup-root string
Optional root cgroup to use for pods. This is handled by the container runtime on a best effort basis. Default: '', which means use the container runtime default. (DEPRECATED: This parameter should be set via the config file specified by the Kubelet's --config flag. See https://kubernetes.io/docs/tasks/administer-cluster/kubelet-config-file/ for more information.)
```

If it uses `--cgroup-root=/container.slice`, then the path could be `/container.slice/kubepods.slice`, i.e. there would be a `container.slice` segment between the `systemd` and the `kubepods` path segments. Currently, there is a `.*` in the regexp to handle that. I will add a unit test for this.

Reply:
Cool. Is there a risk that we will pick up container id and pod uid from an unrelated (not kubepods related) cgroup?

Reply:
I don't see a risk of mistake. `ExtractIdFromCgroupPath()` is called in this context:
- runc prepares some mount related to a cgroup path to prepare the container
- the BPF program detects that activity, and sends a message to the traceloop process via a perf ring buffer
- the traceloop process receives the message about the new container with the cgroup path and the pid of runc as parameters, then it uses `ExtractIdFromCgroupPath()` to figure out if the container comes from a Kubernetes pod.

Both the pod uid and the container id are randomly generated (by kube-apiserver and the container runtime (docker, cri-o) respectively).